include/networkit/graph/Graph.hpp

```diff
@@ -419,11 +422,355 @@ class Graph final {
     }
 
 public:
+    /**
+     * Class to iterate over the nodes of a graph.
+     */
+    class NodeIterator {
+
+        const Graph *G;
+        node u;
+
+        bool inValidState() const noexcept { return G->hasNode(u) || u >= G->upperNodeIdBound(); }
+
+        void nextExistingNode() {
+            do {
+                ++u;
+            } while (!inValidState());
+        }
+
+        void prevExistingNode() {
+            do {
+                --u;
+            } while (!inValidState());
+        }
+
+    public:
+        // The value type of the nodes (i.e. nodes). Returned by
+        // operator*().
+        using value_type = node;
+
+        // Reference to the value_type, required by STL.
+        using reference = value_type &;
+
+        // Pointer to the value_type, required by STL.
+        using pointer = value_type *;
+
+        // STL iterator category.
+        using iterator_category = std::forward_iterator_tag;
+
+        // Signed integer type of the result of subtracting two pointers,
+        // required by STL.
+        using difference_type = ptrdiff_t;
+
+        // Own type.
+        using self = NodeIterator;
+
+        NodeIterator(const Graph *G, node u) : G(G), u(u) {
+            if (!inValidState()) {
+                nextExistingNode();
+            }
+        }
+
+        /**
+         * @brief WARNING: This constructor is required for Python and should not be used as the
+         * iterator is not initialized.
+         */
+        NodeIterator() {}
```

Comment:
Can we put the iterator into a `past-the-end` state so that using it does not involve undefined behavior?

Reply:
Or at least, initialize the graph to `nullptr`.